# CatBoost

In [1]:
from catboost import CatBoostClassifier
import pandas as pd
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

## Importando os dados

In [2]:
Y_treino = pd.read_csv('/home/lucasqueiros/DataScience/Projeto-Classificacao/data/processed/Y_treino.csv', sep=',', encoding='utf-8')
Y_teste = pd.read_csv('/home/lucasqueiros/DataScience/Projeto-Classificacao/data/processed/Y_teste.csv', sep=',', encoding='utf-8')

X_treino = pd.read_csv('/home/lucasqueiros/DataScience/Projeto-Classificacao/data/processed/X_treino.csv', sep=',', encoding='utf-8')
X_teste = pd.read_csv('/home/lucasqueiros/DataScience/Projeto-Classificacao/data/processed/X_teste.csv', sep=',', encoding='utf-8')

In [3]:
Y_teste = Y_teste.iloc[:, 1].values
Y_treino = Y_treino.iloc[:, 1].values

X_esc_teste = X_teste.iloc[:, 1:].values
X_esc_treino = X_treino.iloc[:, 1:].values

### Treinando modelo com dados escalonados

In [4]:
catboost = CatBoostClassifier(task_type='CPU', iterations=100, learning_rate=0.1, depth = 8, random_state = 5,
                              eval_metric="Accuracy")


In [5]:
catboost.fit(X_treino, Y_treino, plot=True, eval_set=(X_teste, Y_teste))

MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

0:	learn: 0.9648241	test: 0.9064327	best: 0.9064327 (0)	total: 110ms	remaining: 10.9s
1:	learn: 0.9773869	test: 0.9415205	best: 0.9415205 (1)	total: 141ms	remaining: 6.91s
2:	learn: 0.9673367	test: 0.9532164	best: 0.9532164 (2)	total: 163ms	remaining: 5.28s
3:	learn: 0.9723618	test: 0.9532164	best: 0.9532164 (2)	total: 188ms	remaining: 4.51s
4:	learn: 0.9824121	test: 0.9590643	best: 0.9590643 (4)	total: 209ms	remaining: 3.97s
5:	learn: 0.9899497	test: 0.9590643	best: 0.9590643 (4)	total: 228ms	remaining: 3.58s
6:	learn: 0.9924623	test: 0.9590643	best: 0.9590643 (4)	total: 253ms	remaining: 3.36s
7:	learn: 0.9874372	test: 0.9590643	best: 0.9590643 (4)	total: 273ms	remaining: 3.14s
8:	learn: 0.9899497	test: 0.9590643	best: 0.9590643 (4)	total: 292ms	remaining: 2.95s
9:	learn: 0.9899497	test: 0.9590643	best: 0.9590643 (4)	total: 314ms	remaining: 2.83s
10:	learn: 0.9899497	test: 0.9590643	best: 0.9590643 (4)	total: 341ms	remaining: 2.76s
11:	learn: 0.9924623	test: 0.9649123	best: 0.9649123 

In [6]:
previsoes_cat = catboost.predict(X_teste)
previsoes_cat

array([0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0,
       1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0,
       1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0,
       0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0])

In [7]:
Y_teste

array([0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0,
       1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0,
       1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0,
       0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0])

In [8]:
print("Acurácia: %.2f%%" % (accuracy_score(Y_teste, previsoes_cat) * 100.0))

Acurácia: 96.49%


## Análise com dados de treino

In [9]:
previsoes_treino = catboost.predict(X_treino)
previsoes_treino

array([0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0,
       1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0,
       1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0,
       1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1,
       1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1,
       0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1,

In [10]:
Y_treino.shape, previsoes_treino.shape

((398,), (398,))

In [11]:
print("Acurácia: %.2f%%" % (accuracy_score(Y_treino, previsoes_treino) * 100.0))

Acurácia: 99.25%


In [12]:
confusion_matrix(Y_treino, previsoes_treino)

array([[249,   0],
       [  3, 146]])

## Validação Cruzada

In [13]:
# Separando os dados em folds
kfold = KFold(n_splits = 30, shuffle=True, random_state = 5)

In [14]:
# Criando o modelo
modelo = CatBoostClassifier(task_type='CPU', iterations=100, learning_rate=0.1, depth = 8, random_state = 5,
                              eval_metric="Accuracy")

In [15]:
resultado = cross_val_score(modelo, X_treino, Y_treino, cv = kfold)
resultado

0:	learn: 0.9557292	total: 27.9ms	remaining: 2.76s
1:	learn: 0.9843750	total: 43.7ms	remaining: 2.14s
2:	learn: 0.9843750	total: 64.9ms	remaining: 2.1s
3:	learn: 0.9843750	total: 82.7ms	remaining: 1.99s
4:	learn: 0.9843750	total: 98.8ms	remaining: 1.88s
5:	learn: 0.9843750	total: 115ms	remaining: 1.8s
6:	learn: 0.9895833	total: 134ms	remaining: 1.77s
7:	learn: 0.9869792	total: 150ms	remaining: 1.72s
8:	learn: 0.9843750	total: 169ms	remaining: 1.7s
9:	learn: 0.9869792	total: 187ms	remaining: 1.68s
10:	learn: 0.9869792	total: 207ms	remaining: 1.68s
11:	learn: 0.9921875	total: 225ms	remaining: 1.65s
12:	learn: 0.9921875	total: 248ms	remaining: 1.66s
13:	learn: 0.9947917	total: 293ms	remaining: 1.8s
14:	learn: 0.9947917	total: 320ms	remaining: 1.81s
15:	learn: 0.9947917	total: 345ms	remaining: 1.81s
16:	learn: 0.9947917	total: 364ms	remaining: 1.78s
17:	learn: 0.9947917	total: 388ms	remaining: 1.77s
18:	learn: 0.9947917	total: 407ms	remaining: 1.74s
19:	learn: 0.9973958	total: 426ms	remain

array([0.78571429, 1.        , 1.        , 0.92857143, 1.        ,
       0.92857143, 1.        , 1.        , 1.        , 0.92307692,
       1.        , 1.        , 0.92307692, 1.        , 1.        ,
       1.        , 1.        , 1.        , 1.        , 1.        ,
       0.92307692, 0.92307692, 1.        , 0.92307692, 1.        ,
       1.        , 0.84615385, 0.76923077, 1.        , 1.        ])

In [16]:
# Usamos a média e o desvio padrão
print("Acurácia Média: %.2f%%" % (resultado.mean() * 100.0))

Acurácia Média: 96.25%


##### CatBoost = 96,49% (treino e teste) e 96.25% (validação cruzada) -  previsores - CatBoostClassifier(task_type='CPU', iterations=100, learning_rate=0.1, depth = 8, random_state = 5,eval_metric="Accuracy")